### PR TITLE
SWA over cosine tail (ep46-50) on 5L/256d baseline

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Stochastic Weight Averaging (SWA) over the cosine tail (ep40–50) will reduce wsy/wsz variance and improve `abupt_axis_mean_rel_l2_pct` vs a single-checkpoint model.**

The current best recipe (5L/256d + FourierPE + T_max=50) produces a best checkpoint at ep~45.3 (step 807,025). However, the cosine schedule flattens in the tail (ep40–50), meaning multiple checkpoints in this range represent similarly-trained weights on a near-flat loss landscape. SWA averages checkpoints from this flat tail to find a wider, flatter minimum — a well-established technique shown to improve generalization across multiple domains (Izmailov et al. 2018, "Averaging Weights Leads to Wider Optima and Better Generalization").

The key intuition: at ep40–50, learning rate is ~0 and checkpoints oscillate in a tight basin around the loss minimum. Averaging these checkpoints is equivalent to implicit ensemble regularization that smooths weight-space curvature. For a CFD surrogate, the wsy/wsz channels carry the highest per-sample variance; SWA should disproportionately benefit the hardest channels.

**Approach:** Train to ep50 normally (no changes to the training loop). After training, compute a SWA checkpoint by averaging the model weights from the last 5 saved checkpoints (ep46–50, or equivalently the 5 best by val_primary during the tail). Evaluate the SWA checkpoint on val_primary and compare to the standard best-checkpoint result.

## Instructions

Base config: **5L/256d/4H + FourierPE + T_max=50 + no-EMA + lr=3e-4** (the current best recipe from PR #174).

**Step 1: Run standard training to ep50**

```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --no-use-ema \
  --lr 3e-4 \
  --epochs 50 \
  --lr-cosine-t-max 50 \
  --lr-min 1e-6 \
  --lr-warmup-epochs 3 \
  --no-compile-model \
  --kill-thresholds "500:train/loss<5,100000:val_primary/abupt_axis_mean_rel_l2_pct<25" \
  --wandb-group bengio-wave12-swa \
  --wandb-name kohaku/swa-cosine-tail \
  --agent kohaku
```

**Step 2: After training, compute SWA checkpoint**

After training completes, locate the saved checkpoints for epochs 46–50 (the cosine tail). Average the model state_dicts from these 5 checkpoints:

```python
import torch
import glob
import os

# Load checkpoints from cosine tail (ep46-50)
ckpt_paths = sorted(glob.glob("checkpoints/kohaku/swa-cosine-tail/ep4[6-9]*.pt") + 
                    glob.glob("checkpoints/kohaku/swa-cosine-tail/ep50*.pt"))
print(f"SWA averaging {len(ckpt_paths)} checkpoints: {ckpt_paths}")

# Average weights
avg_state = None
for p in ckpt_paths:
    ckpt = torch.load(p, map_location='cpu')
    state = ckpt['model_state_dict']
    if avg_state is None:
        avg_state = {k: v.clone().float() for k, v in state.items()}
    else:
        for k in avg_state:
            avg_state[k] += state[k].float()

for k in avg_state:
    avg_state[k] /= len(ckpt_paths)

# Save SWA checkpoint (reuse one original ckpt as template)
template = torch.load(ckpt_paths[-1], map_location='cpu')
template['model_state_dict'] = avg_state
torch.save(template, "checkpoints/kohaku/swa-cosine-tail/swa_ep46-50.pt")
print("Saved SWA checkpoint.")
```

**Step 3: Evaluate SWA checkpoint**

Run validation on the SWA checkpoint and report full `val_primary` metrics. Compare against the best single-checkpoint result from Step 1.

Adapt the evaluation command to point at the SWA checkpoint file. If `train.py` has an `--eval-only` or `--resume` flag, use that. Otherwise implement a minimal eval loop using the model architecture from `model.py`.

**Checkpoint naming convention:** if the training saves checkpoints differently (by step or by W&B run ID), adapt the glob pattern accordingly. Print the full list of available checkpoint files early and adjust.

## Baseline

Current best (PR #174, W&B run `vu4jsiic`, ep~45.3, step 807,025):

| Metric | Best val |
|--------|---------|
| `abupt_axis_mean_rel_l2_pct` | **6.9549** |
| `surface_pressure_rel_l2_pct` | 4.5644 |
| `wall_shear_y_rel_l2_pct` | 8.7345 |
| `wall_shear_z_rel_l2_pct` | 10.5766 |
| `volume_pressure_rel_l2_pct` | 3.9361 |

Target to beat: **val abupt < 6.9549%** (either from the best single checkpoint during training, or from the SWA checkpoint)

## Expected outcome

SWA averaging the last 5 checkpoints should either:
- Match the best single checkpoint (if the basin is already flat), confirming stability
- Improve abupt by −0.1 to −0.5pp by finding a wider minimum

The per-axis improvement on wsy and wsz relative to sp is the key signal to watch. If SWA preferentially reduces wsy/wsz at the cost of sp, it is still a net positive for abupt.

## Reporting

1. After training reaches ep46+, report the per-epoch trajectory for the last 5 epochs with full `val_primary` breakdown.
2. After SWA eval, report the SWA vs best-single-checkpoint comparison in a table.
3. Include the W&B run ID in your first update comment.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Acknowledged — hypothesis received. SWA over the cosine tail (ep46–50) on 5L/256d baseline is a low-risk post-training smoothing technique that can sometimes squeeze out +0.1–0.2pp. Current baseline to beat: **val_abupt = 6.9549%** (PR #174, run `vu4jsiic`, ep~45.3).

Note: SWA requires training to at least ep46 first, then averaging weights over ep46–50. Make sure you start from or reproduce the PR #174 training setup (5L/256d/4H, FourierPE, no-EMA, lr=3e-4, T_max=50) so the comparison is fair.

Report: (1) last-epoch (ep50) val_abupt without SWA, (2) SWA-averaged val_abupt, (3) per-axis breakdown for both.

Post your W&B run ID(s) and first val checkpoint when training starts.

**Comment by morganmcg1** [comment]
STUDENT kohaku:

Started SWA training run.

## Setup

- **W&B run (rank 0, canonical):** [5v1mjka1](https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml/runs/5v1mjka1)
- **DDP rank runs:** 5v1mjka1 (r0), 5dywhv3l (r1), mhn1y0wx (r2), j4wjk8ie (r3)
- **Group:** `bengio-wave12-swa`
- **Throughput:** ~8.87 it/s on 4×RTX PRO 6000 Blackwell, ~33 min/epoch, ~28h total wall-clock estimate
- **Steps/epoch:** 17,816 (matches PR #174's vu4jsiic at 807,025/45.3 ep ≈ 17,816)

## Implementation notes

I added two flags to `train.py` to wire SWA inline (not as a separate offline script) so the SWA eval reuses the existing distributed/full-fidelity eval infrastructure and logs metrics to the same W&B run:

- `--swa-tail-epochs 10` → saves rank-0 checkpoints for ep41–50 in `outputs/.../run-<id>/swa_tail_ckpts/epoch_NNN.pt`. Same checkpoint format as the existing `checkpoint.pt`.
- `--swa-eval-ks 5,10` → after the standard `run_final_evaluation` finishes, averages the last K checkpoints uniformly and runs full-fidelity `full_val` + `test` eval on each, logging under `swa{k}_full_val_primary/*` and `swa{k}_test_primary/*`.

I'm reporting **K=5 (ep46–50) per the PR instructions** and **K=10 (ep41–50) as a comparison** based on a research pass:
- LAWA / Model Soups suggest k=5–10 is the sweet spot
- 0.2pp on val_abupt is the minimum effect threshold for "real" (research agent estimate); will not claim a win below that bar

## Research pass key findings

- **Equal weights are right** (Izmailov 2018, Wortsman 2022 Model Soups)
- **No buffer recomputation needed** — LayerNorm / RMSNorm have no running statistics buffers; the model only has parameter weights to average
- **This is tail-averaging not true SWA** (Guo et al. 2022 "SWA Revisited" at arXiv 2201.00519 distinguishes the two; true SWA needs a constant-LR cyclic phase to explore the basin). Cosine-decayed-tail averaging acts more like noise reduction within an already-narrow minimum, so the expected gain is 0.05–0.3pp, not the 1–2pp basin-geometry win
- **No published prior for SWA on neural operators / CFD surrogates** as of May 2026 — we are in uncharted territory here

## PR command deviation note

The PR command has `--lr-warmup-epochs 3`, but the BASELINE.md PR #174 reproduce command does **not** set warmup (default 0). This means the LR schedule differs slightly: with warmup=3, cosine starts at ep4 and decays to ~3.7e-6 at ep50 (vs. ~1e-6 at ep50 without warmup). I'm following the PR command exactly as written; if the standard best-checkpoint val_abupt comes in noticeably worse than the 6.9549% baseline, this warmup deviation is the most likely cause.

Will update with full per-epoch trajectory + SWA comparison once training finishes.

**Comment by morganmcg1** [comment]
ADVISOR (deviation flag): Your reproduce command includes `--lr-warmup-epochs 3`, which is NOT in the canonical baseline (PR #174 / BASELINE.md). This causes cosine to decay to ~3.7e-6 at ep50 instead of ~1e-6 without warmup, slightly altering the final-epoch checkpoint quality.

For SWA-tail comparison fidelity, this is acceptable since SWA averages ep46-50 and the absolute LR floor matters less than the schedule curvature. But please **note the deviation in your results write-up** so we can attribute any delta correctly. Continue the run; do not relaunch.
